plugins/modules/mso_service_node_type.py

```diff
@@ -0,0 +1,159 @@
+
+- name: Remove a Service Node Type
+  cisco.mso.mso_schema_service_node:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    name: Firewall
+    state: absent
+  delegate_to: localhost
+
+- name: Query a specific Service Node Type
+  cisco.mso.mso_schema_service_node:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    name: Firewall
+    state: query
+  delegate_to: localhost
+
+- name: Query all Service Node Types
+  cisco.mso.mso_schema_service_node:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    state: query
+  delegate_to: localhost
+'''
+
+RETURN = r'''
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cisco.mso.plugins.module_utils.mso import MSOModule, mso_argument_spec
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        name=dict(type='str'),
+        display_name=dict(type='str'),
+        state=dict(type='str', default='present', choices=['absent', 'present', 'query']),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ['state', 'absent', ['name']],
+            ['state', 'present', ['name']],
+        ],
+    )
+
+    name = module.params.get('name')
+    display_name = module.params.get('display_name')
+    state = module.params.get('state')
+
+    mso = MSOModule(module)
+
+    mso.existing = {}
+    service_node_id = None
+
+    # Get service node id
+    query_node_data = mso.query_service_node_types()
+    service_nodes = [f.get('name') for f in query_node_data]
+    if name in service_nodes:
+        for node_data in query_node_data:
+            if node_data.get('name') == name:
+                service_node_id = node_data.get('id')
+                mso.existing = node_data
+
+    if state == 'query':
+        if name is None:
+            mso.existing = query_node_data
+        if name is not None and service_node_id is None:
+            mso.fail_json(msg="Service Node Type '{service_node_type}' not found".format(service_node_type=name))
+        mso.exit_json()
+
+    service_nodes_path = '/schemas/service-node-types'
+    service_node_path = '/schemas/service-node-types/{0}'.format(service_node_id)
+
+    mso.previous = mso.existing
+    if state == 'absent':
+        if mso.existing:
+            if module.check_mode:
+                mso.existing = {}
+            else:
+                mso.existing = mso.request(service_node_path, method='DELETE')
+
+    elif state == 'present':
+        if display_name is None:
+            display_name = name
+
+        payload = dict(
+            name=name,
+            displayName=display_name,
+        )
+        mso.sanitize(payload, collate=True)
+        if not module.check_mode:
+            if not mso.existing:
+                mso.request(service_nodes_path, method='POST', data=payload)
+            else:
```

Comment:
Don't fail if exist but identical to what is provided by the user. Only if different because you cannot update it.